### PR TITLE
feat: initial implementation of protoc-gen-go-aip-lint

### PR DESCRIPTION
**Pull request by odsod**
- feat: initial implementation of protoc-gen-go-aip-lint
- fix: update aipmiddleware.ParentValidator to conform to AIP-4231

Lint rules that complement the official [API Linter][api-linter] with
additional rules.

[api-linter]: https://linter.aip.dev/

### CODE REVIEW

File: buf.gen.yaml [hunk 1]
```diff
@@ -4,6 +4,9 @@ plugins:
   - name: gapic-validator
     out: .
 
+  - name: go-aip-lint
```
Comment (ericwenn) [inline buf.gen.yaml:7]: Is the linter specific for `go` or should it just be called `aip-lint` (`protoc-gen-aip-lint`).

**Review by ericwenn — APPROVED** [review]
How do we handle tests for the linter? For example `api-linter` has go-test coverage for most rules IIRC.

**Comment by odsod** [comment]
@ericwenn I am working on a test harness for the api-linter that does not depend on jhump/protoreflect - stay tuned!

File: buf.gen.yaml (hunk 1 above)
Comment (odsod) [reply buf.gen.yaml:7]: I'm thinking that the scope of the linter is that it should ensure that the upcoming code generator in `aip-go` will work well assuming the linter passes - so it's a linter specifically for `aip-go` - if that makes sense?

Comment (ericwenn) [reply buf.gen.yaml:7]: makes sense, `protoc-gen-aip-lint` could be a bit of a broad name as well.

**Comment by ericwenn** [comment]
@odsod close?

**Pull request closed without merging**